crates/red_knot_python_semantic/src/symbol.rs

```diff
@@ -44,15 +44,25 @@ impl<'db> Symbol<'db> {
         }
     }
 
-    pub(crate) fn unwrap_or(&self, other: Type<'db>) -> Type<'db> {
+    pub(crate) fn unwrap_or_unknown(&self) -> Type<'db> {
         match self {
             Symbol::Type(ty, _) => *ty,
-            Symbol::Unbound => other,
+            Symbol::Unbound => Type::Unknown,
         }
     }
 
-    pub(crate) fn unwrap_or_unknown(&self) -> Type<'db> {
-        self.unwrap_or(Type::Unknown)
+    /// Get the return type of a call to a symbol if it is definitely bound and callable.
+    ///
+    /// Returns `None` if the symbol is (possibly) unbound or if the type is not callable.
+    pub(crate) fn get_return_type_if_callable(
```

Comment:
I would just call this return_type_if_callable. Getters are rather uncommon in Rust (set methods are more common but also rear)

Reply:
I had `return_type_if_callable` first, but it could be read as: "return the (inner) type if this symbol is callable" — which is not what it does :man_shrugging:. Due to the dual meaning of "return" here. But I'm happy to select the shorter name.

Reply:
Hmm, I see. How about `callable_return_type` : The callable's return type`

Reply:
Alternative suggestion: just name it `call` :) This is parallel to `Type::call`, and I think it's a reasonable assumption that the return `Type` of a method named `call` is the type returned by the call.

I also think we could improve explicitness and consistency here significantly, while not regressing ergonomics much, if this returned a `CallOutcome` and we added `MaybeNotBound` as a possible `CallOutcome`? Then we can continue to consolidate and improve our APIs on `CallOutcome` (which already includes a method for "get me the return type as Option<Type> or None if error") and consistently return those from all calls. Then call sites which really do want to ignore all reasons for the error can do so with just one more chained method call, but it's a bit more explicit what they are doing?

Reply:
> I also think we could improve explicitness and consistency here significantly, while not regressing ergonomics much, if this returned a `CallOutcome` and we added `MaybeNotBound` as a possible `CallOutcome`?

I had tried this before, and did so again now. I think it's more explicit, but the disadvantage of this approach is that we now have a few `CallOutcome` variants that are possible results of `Symbol::call`, but not of `Type::call`. Callers which use `Type::call` will now have to (implicitly or explicitly) ignore the possibility of `CallOutcome::Unbound` and similar. This is not a major issue right now, though. A related discussion is #13996, because we now squeeze even more cases through `return_ty` (*what does `None` mean?*) and `return_ty_result` (*under which circumstances will I get a return type?*).

Reply:
On the plus side, this new version helped me to find some previously unhandled cases of possible unboundness. See new tests in `call/callable_instance.md`